File: LeetCode/Monthly Chalange/February-23/103. Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.zig = []
        
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        self.dfs(root, 1, 0)
        for i in range (1, len(self.zig), 2):
            self.zig[i].reverse()
        return self.zig
    
    def dfs(self, root: Optional[TreeNode],f:bool, lv: int) -> None:
        if not root:
            return
        if lv>=len(self.zig):
            self.zig.append([])
        
        self.zig[lv].append(root.val)
        
        if root.left:
            self.dfs(root.left, not f, lv+1)
        if root.right:
            self.dfs(root.right, not f, lv+1)
        

File: LeetCode/Monthly Chalange/February-23/103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zig;
    
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        dfs(root, 1, 0);
        for(int i=1;i<zig.size();i+=2) reverse(zig[i].begin(), zig[i].end()); 
        return zig;
    }
    
private:
    void dfs(TreeNode* start, int f, int lv){
        if(start == NULL) return;
        if(lv >= zig.size()) zig.push_back(vector<int>());
        
        zig[lv].push_back(start->val);

        if (start->left != NULL) dfs(start->left, 1-f, lv+1);
        if (start->right != NULL) dfs(start->right, 1-f, lv+1);
    }
};
